internal/topic/topicreaderinternal/stream_reader_impl.go

```diff
@@ -185,21 +185,43 @@ func (r *topicStreamReaderImpl) WaitInit(_ context.Context) error {
 	return nil
 }
 
+func (r *topicStreamReaderImpl) TopicOnReaderStart(consumer string, err error) {
+	onReaderStart := func(ctx context.Context) {
+		trace.TopicOnReaderStart(r.cfg.Trace, &ctx, r.readerID, consumer, err)
+	}
+	r.traceWithLogContext(onReaderStart)
+}
+
+type traceFunction func(ctx context.Context)
+
+func (r *topicStreamReaderImpl) traceWithLogContext(f traceFunction) {
+	r.m.WithLock(func() {
+		logCtx := r.cfg.LogContext
+		f(logCtx)
+		r.cfg.LogContext = logCtx
+	})
+}
+
 func (r *topicStreamReaderImpl) PopMessagesBatchTx(
 	ctx context.Context,
 	tx tx.Transaction,
 	opts ReadMessageBatchOptions,
 ) (_ *topicreadercommon.PublicBatch, resErr error) {
-	traceCtx := ctx
-	onDone := trace.TopicOnReaderStreamPopBatchTx(
-		r.cfg.Trace,
-		&traceCtx,
-		r.readerID,
-		r.readConnectionID,
-		tx.SessionID(),
-		tx,
-	)
-	ctx = traceCtx
+	onDone := func(err error) {
+		r.m.WithLock(func() {
+			logCtx := r.cfg.LogContext
+			trace.TopicOnReaderStreamPopBatchTx(
+				r.cfg.Trace,
+				&logCtx,
+				r.readerID,
+				r.readConnectionID,
+				tx.SessionID(),
+				tx,
+			)(err)
+			r.cfg.LogContext = logCtx
```

Comment:
1. a config is not storage, the content of config should not changed after receive.
2. the context field rewrite in many placed without sync - it is good way for data races
3. The context infinite accumulate any additional fields from traces - it is good way for memory leak